ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -765,6 +805,67 @@ module Winbind = struct
       hostname
 end
 
+module ClosestKdc = struct
+  let periodic_update_task_name = "Update closest kdc"
+  let startup_delay = 5.
+
+  let mtime_this ~name ~f =
+    let start = Mtime_clock.counter() in
+    match f() with
+    | Ok _ -> Ok(name, Mtime_clock.count start)
+    | Error e -> Error e
+
+  let update_db ~domain ~kdc =
+  (fun __context ->
+     let self = Helpers.get_localhost ~__context in
+     Db.Host.get_external_auth_configuration ~__context ~self|> fun value ->
+     (domain, kdc) :: List.remove_assoc domain value |> fun value ->
+     Db.Host.set_external_auth_configuration ~__context ~self ~value)
+  |> Server_helpers.exec_with_new_task "update domain closest kdc"
+
+  let from_db domain =
+    (fun __context ->
+     let self = Helpers.get_localhost ~__context in
+     Db.Host.get_external_auth_configuration ~__context ~self|> List.assoc_opt domain)
+  |> Server_helpers.exec_with_new_task "query domain closest kdc"
+
+  let lookup domain =
+    try
+      let* krbtgt_sid = Wbinfo.sid_of_name (Printf.sprintf "%s@%s" krbtgt domain) in
+      let* domain_netbios_name = Wbinfo.domain_name_of ~target_name_type:NetbiosName ~from_name: domain in
+
+      kdcs_of_domain domain
+      |> List.map (fun kdc -> debug "Got domain '%s' kdc '%s'" domain kdc; kdc)
+      |> List.map (fun kdc -> mtime_this ~name:kdc ~f:(fun() -> Ldap.query_user krbtgt_sid domain_netbios_name kdc))
```

Comment:
If one of the KDCs is down this will fail, however the admin can just do a toolstack restart once they fixed the KDC, so this should be fine. If needed we could ignore failures here and use filter_map

Reply:
If we find the closest DC, we will trust it and use it for the next whole day, 
If the chosen DC is down, yes, this issue can be simply resolved by xapi restart.
Also, this check interval can be configured in xapi configure file. (we do not expected the DC fail often)